programmer/tinyprog/__init__.py

```diff
@@ -376,6 +379,84 @@ def program(self, addr, data):
             self.progress("Failure!")
             return False
 
+    # for each sector: erase,program,verify
+    # this programming is much more reliable than previous program()
+    # because each written sector will be verified and in case of
+    # error, retried several times
+    def program_sectors(self, addr, data, retry=10):
+        possible_lengths = (1, 4 * 1024, 32 * 1024, 64 * 1024)
+        retries_remaining = retry
+        length = len(data)
+        offset = 0
+        write_addr = addr + offset
+        with tqdm(desc="    Writing", unit="B", unit_scale=True, total=length, disable=False) as pbar:
+            while length > 0 and retries_remaining > 0:
```

Comment:
This entire while loop was copied and pasted from the `erase` function...instead of copy and pasting the entire body of the function, why not call the function from here instead?  I can't accept the same large block of code duplicated three times.

Reply:
Yes actually the erase function was used as basis of sector program function.
I'm not a good python programmer, I needed quick fix just to make it work for me.
If someone can cleaner rewrite everything from scratch, its OK for me just
follow the logic: each sector should be erased, written and verified with multiple
retrieed (10x) for successful upload of 1-2MB or longer data 

I did similar bootloadings with sectors for AVR with unreliable
data link so I know what and how should be done.

Of course, if USB compatibility part is fixed to not loose packets,
then everyting will work even better. I suggest moving to HID packets
instead of trying to emulate usb-serial.

Reply:
Whether it's HID or serial doesn't actually make a difference for the reliability.  In the end it's still two bulk endpoints transfering data.  There is likely a bug in the code, I have noticed some specific data patterns that will cause the issue, but I haven't been able to track it down.  I can see if I can't write up a smaller algorithm to do the retries.

How big of a sector are you using? 1MByte?

Reply:
There is (I guess) important but invisilble difference. Serial transfer
need to be bidirectional and each direction needs to be a part of continuous
data stream. My theory is that in verilog code of simplified USB protocol this
can't be properly handing without introducing much complexity with semaphores
buffering mutexing or similar. Simultaneously running bidirectional strems poses
additional challenge for potentioanly as-simple-as possible USB core.

HID doesn't treat data as streams on core level but as packets so it's more closer to UDP.
Assembly of the bidirectionaly stream will be left to python which is much more convenient
to be done than in verilog.

I'm using sector size same as erase size because the funciton is modified erase function.
so sector is most of the times 64K but at the end of data it usually becomes shorter.